openslides_backend/presenter/base.py

```diff
@@ -1,19 +1,43 @@
-from typing import Any, Dict
+from typing import Any, Dict, Optional, Callable
+from ..shared.exceptions import PresenterException
 
+from ..services.datastore.interface import Datastore
+from ..shared.interfaces import Permission, LoggingModule
+from fastjsonschema import JsonSchemaException
 
-class PresenterBase:  # pragma: no cover
+
+class BasePresenter:
     """
-    Abstract base class for presenters.
+    Base class for presenters.
     """
 
-    @property
-    def data(self) -> Dict[Any, Any]:
-        ...
-
+    data: Optional[Any]
+    permission: Permission
+    database: Datastore
+    logging: LoggingModule
 
-class Presenter(PresenterBase):
-    """
-    Base clase for presenters.
-    """
+    def __init__(self, data: Optional[Any], permission: Permission, datastore: Datastore, logging: LoggingModule):
+        self.data = data
+        self.permission = permission
+        self.datastore = datastore
+        self.logging = logging
+        self.logger = logging.getLogger(__name__)
+    
+    def validate(self) -> None:
+        """ Validates the given data. If self.schema is not set, assumes that no data should be given. """
+        if self.schema:
+            if self.data is None:
+                raise PresenterException("No data given.")
+            try:
+                # unfortunately, python injects self if we use the bounded method, so
+                # we have to use the class method
+                self.__class__.schema(self.data)
```

Comment:
In BaseAction I use `type(self).schema(...)`. Which solution is better?

Reply:
Ahh, didn't see that, that would have saved me quite some trouble... I think yours is cleaner, will change it, thanks